File: server/server-core/src/main/java/com/simple/castle/server/manager/SceneManager.java
package com.simple.castle.server.manager;

import com.badlogic.gdx.graphics.g3d.Model;
import com.badlogic.gdx.graphics.g3d.ModelInstance;
import com.badlogic.gdx.utils.Disposable;
import com.simple.castle.base.ModelSend;
import com.simple.castle.server.composition.BaseObject;
import com.simple.castle.server.composition.Constructor;
import com.simple.castle.server.json.SceneObjectsJson;

import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

public class SceneManager implements Disposable {
    private final Map<String, Constructor> constructorMap;
    private final Map<String, BaseObject> baseObjectMap;
    private final List<ModelInstance> drawables;
    private final List<ModelSend> modelsSend;

    public SceneManager(SceneObjectsJson sceneObjectsJson, Model model) {
        constructorMap = sceneObjectsJson.getSceneObjectsJson()
                .stream()
                .map(sceneObjectJson -> new Constructor(model,
                        sceneObjectJson.getNodePattern(),
                        sceneObjectJson.getInteract(),
                        sceneObjectJson.getShape(),
                        sceneObjectJson.getInstantiate(),
                        sceneObjectJson.getHide()))
                .collect(Collectors.toMap(Constructor::getId, Function.identity()));

        baseObjectMap = constructorMap.entrySet()
                .stream()
                .filter(entry -> entry.getValue().getInstantiate())
                .collect(Collectors.toMap(Map.Entry::getKey, entry -> new BaseObject(entry.getValue())));

        drawables = baseObjectMap.values().stream()
                .filter(baseObject -> !baseObject.getHide())
                .map(BaseObject::getModelInstance).collect(Collectors.toList());

        modelsSend = baseObjectMap.values().stream()
                .filter(baseObject -> !baseObject.getHide())
                .map(baseObject -> {
                    ModelSend modelSend = new ModelSend();
                    modelSend.setId(baseObject.getId());
                    modelSend.setMatrix4(baseObject.getModelInstance().transform);
                    return modelSend;
                }).collect(Collectors.toList());
    }

    public BaseObject getObject(String id) {
        return baseObjectMap.get(id);
    }

    public List<ModelInstance> getDrawables() {
        return drawables;
    }

    public Collection<BaseObject> getAll() {
        return baseObjectMap.values();
    }

    public List<ModelSend> getSend() {
        return modelsSend;
    }

    @Override
    public void dispose() {
        baseObjectMap.values().forEach(BaseObject::dispose);
    }
}

File: server/server-core/src/main/java/com/simple/castle/server/kt/manager/SceneManager.kt
package com.simple.castle.server.kt.manager

import com.badlogic.gdx.graphics.g3d.Model
import com.badlogic.gdx.graphics.g3d.ModelInstance
import com.badlogic.gdx.utils.Disposable
import com.simple.castle.server.kt.composition.BaseObject
import com.simple.castle.server.kt.composition.Constructor
import com.simple.castle.server.kt.loader.json.SceneObjectsJson

class SceneManager(sceneObjectsJson: SceneObjectsJson,
                   model: Model?) : Disposable {
    private val constructorMap: Map<String, Constructor>
    private val baseObjectMap: Map<String, BaseObject>
    val drawables: List<ModelInstance>

    val all: Collection<BaseObject>
        get() = baseObjectMap.values

    init {
        constructorMap = sceneObjectsJson.sceneObjectsJson
                ?.map { sceneObjectJson ->
                    Constructor(model!!,
                            sceneObjectJson!!.nodePattern!!,
                            sceneObjectJson.interact!!,
                            sceneObjectJson.shape!!,
                            sceneObjectJson.instantiate!!,
                            sceneObjectJson.hide!!)
                }
                ?.associateBy(keySelector = { constructor -> constructor.id }, valueTransform = { constructor -> constructor })!!

        baseObjectMap = constructorMap.entries
                .filter { entry -> entry.value.instantiate }
                .associateBy(keySelector = { entry -> entry.key }, valueTransform = { entry -> BaseObject(entry.value) })

        drawables = baseObjectMap.values
                .filter { baseObject -> !baseObject.hide!! }
                .map { baseObject -> baseObject.modelInstance!! }

    }

    fun getObject(id: String?): BaseObject? {
        return baseObjectMap[id]
    }

    override fun dispose() {
        baseObjectMap.values.forEach { obj: BaseObject -> obj.dispose() }
    }
}